api/src/main/proto/stellarstation/api/v1/radio/radio.proto

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2018 Infostellar, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+syntax = "proto3";
+
+package stellarstation.api.v1.radio;
+
+option go_package = "radio";
+
+option java_multiple_files = true;
+option java_outer_classname = "RadioProto";
+option java_package = "com.stellarstation.api.v1.radio";
+
+// Messages for modeling radio devices (e.g., receivers, transmitters) in API messages.
+
+// A radio device for RF communication. Usually, a satellite will have one or more radio devices for
+// communicating with the ground and a ground station will have one or more radio devices for
+// communicating with the satellite. Such devices may not actually correspond to actual hardware
+// devices, i.e., in the case of Software Defined Radio (SDR).
+message RadioDevice {
+  // The center frequency of the device, in Hz.
+  uint64 center_frequency_hz = 1;
+
+  // The protocol used by this device when doing RF communication.
+  CommunicationProtocol protocol = 2;
+}
+
+// A communication protocol used with a radio device. These must contain all the parameters
+// needed to configure the device for use either in transmission or reception.
+message CommunicationProtocol {
```

Comment:
Are we going to handle the special cases like APT and CW here (that define both modulation and encoding in their standard)? 

Also, for higher bands, I think it's common to use a combination of error correction algorithms like Viterbi and Reed Solomon, not just one.

I guess we could handle those moving forward based on need.

Reply:
Yeah I think we can add them later as needed, I think it'll just be comment tweaking (e.g., modulation not filled for APT).

Reply:
In this structure, Modulation is a parameter for the radio configuration - we should move it from here to the radio configuration device.

Reply:
From what I understand, there are communication protocols which define both framing and modulation as part of the standard (e.g., APT). In that case I think it's appropriate to keep them together, and later we'll have cases where modulation isn't necessarily provided since it's implied by the framing.

Reply:
First of all, sorry for the long reply, but I think that it is required that I properly explained all the things that are mixed in this discussion.

NOAA weather satellites are the main users of this APT protocol (https://en.wikipedia.org/wiki/Automatic_picture_transmission) although others like METEOR also implement it. The implementation for their satellites uses a specific modulation and bitrate, but this is just the implementation that they have chosen. Since nobody else was doing anything similar, people adopted these modulations and bitrates. However, you can use the APT framing format with a different set of modulations and bitrates; something that it is not done since the satellites and the ground equipment is now compatible with that whole data set.

This is something that happens sometimes in the telecommunications industry: certain protocols are used only with a set of underlaying radio devices because their application is very specific. In the end, the whole set kind of becomes a standard.

In our case, we should separate protocol from modulations and bitrates as a general rule. For example, AX.25 can be used with a broad range of modulations and bitrates.

For special cases like APT, I recommend labeling it as "APT/NOAA" or something alike; but always referring to the fact that we are using NOAA convention for the modulation, bitrate, polarization and central frequency. In this case, if the user selects the "APT/NOAA" protocol in the communications API, the parameters of the RadioConfiguration device should be set accordingly and users cannot modify them.

We might set a different protocol called "APT" where we use the framing as defined by NOAA, but we freely allow users to freely change the RadioConfiguration device parameters.

In addition to this, please also note that CW is a modulation and not a protocol (please refer to: https://en.wikipedia.org/wiki/Continuous_wave). On top of it, for beaconing from satellites, there is a data format that is widely used specially within the CubeSat community to just broadcast the codename for the satellite together with the battery voltage. But this is done on top of CW, that it is a modulation.

For example, in the case of the Birds satellites, their message in the beacon follows this format: 

http://birds2.birds-project.com/operation/

You can tell that they are using the same modulation (CW), but the frame format differs from others like: 

http://www.dk3wn.info/p/?cat=90

This was the beacon for our first satellite in Vigo called XaTcobeo (re-entered the atmosphere years ago).

Hope this helps - let me know if you wanna further discuss about this in person, it is easier most of the times! :)

Reply:
I think this means modulation is independent from protocol so it should be separate. That makes sense, actually it means I don't need a BITSTREAM type of protocol anymore since the lack of a protocol is BITSTREAM.

Reply:
LGTM